helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -202,126 +222,188 @@ public void disconnect() {
+    }
+
+    // Deserialize the record to retrieve the original
+    ZNRecord originalRecord = (ZNRecord) _zkSerializer.deserialize(serializedRecord);
+    return new HelixProperty(originalRecord);
+  }
+
+  @Override
+  public void close() {
+    disconnect();
   }
 
   /**
-   * Returns the last string element in a split String array by /.
+   * Deletes all versions that are older than the currentVersion given.
    * @param path
-   * @return
+   * @param currentVersion
    */
-  private String extractIdFromPath(String path) {
-    String[] splitPath = path.split("/");
-    return splitPath[splitPath.length - 1];
+  private void deleteStaleVersions(String path, long currentVersion) {
+    // Get all children names under path
+    List<String> children = _zkBaseDataAccessor.getChildNames(path, AccessOption.PERSISTENT);
+    if (children == null || children.isEmpty()) {
+      // The whole path has been deleted so return immediately
+      return;
+    }
+    filterChildrenNames(children, currentVersion);
+    List<String> pathsToDelete = getPathsToDelete(path, children);
+    for (String pathToDelete : pathsToDelete) {
+      // TODO: Should be batch delete but it doesn't work. It's okay since this runs async
+      _zkBaseDataAccessor.remove(pathToDelete, AccessOption.PERSISTENT);
+    }
   }
 
   /**
-   * Acquires the lock (create an ephemeral node) only if it is free (no ephemeral node already
-   * exists) at the time of invocation.
-   * @param path
+   * Filter out currentVersion and non-version children names from the children list given.
+   * @param children
+   * @param currentVersion
+   */
+  private void filterChildrenNames(List<String> children, long currentVersion) {
+    // Leave out metadata
+    children.remove(LAST_SUCCESSFUL_WRITE_KEY);
+    children.remove(LAST_WRITE_KEY);
+
+    // Keep the children that are not expired and exclude currentVersion from deletion
+    long currTime = System.currentTimeMillis();
+    children.removeIf(childName -> currTime < extractTimestamp(childName) + _versionTTL
+        || extractVersion(childName) >= currentVersion);
+  }
+
+  /**
+   * Extracts timestamp from the znode name.
+   * @param name
    * @return
    */
-  private boolean tryLock(String path) {
-    // Check if another write is taking place and if not, create an ephemeral node to simulate
-    // acquiring of a lock
-    return !_zkBaseDataAccessor.exists(path + "/" + WRITE_LOCK_KEY, AccessOption.EPHEMERAL)
-        && _zkBaseDataAccessor.set(path + "/" + WRITE_LOCK_KEY, new byte[0],
-            AccessOption.EPHEMERAL);
+  private long extractTimestamp(String name) {
+    // Path structure: ~~/ver_timestamp
+    String[] parts = name.split("_");
+    if (parts.length == 2) {
+      String timestampStr = parts[1];
+      long timestamp;
+      try {
+        timestamp = Long.parseLong(timestampStr);
+        return timestamp;
+      } catch (Exception e) {
+        LOG.warn(
```

Comment:
If one random node exists, I guess this warning will flood our log file. Shall we do a debug log here?